File: jib.py
"""
Provides all functions to build the jib of a crane
"""

import numpy as np

nodes = []
beams = []


class Dims:
    """Class that contains all dimensions of the jib for global control"""
    SEGMENT_LENGTH = 0
    SEGMENTS = 0
    HEIGHT = 0
    
    START_HEIGHT = 0
    TOWER_WIDTH = 0
    INIT_BAR = 0
    
    IS_CONNECTED = False
    
    TOTAL_LENGTH = 0


def create(tower_height, tower_width, length, segments):
    """
    Creates jib separately from the tower but with the ability to connect to it

    Args:
    :param tower_height: height of the tower (float)
    :param tower_width: width of the tower (float)
    :param length: length of the jib to be created  (float)
    :param segments: number of segments the jib should consist of  (int)
    """
    Dims.START_HEIGHT = tower_height
    Dims.TOWER_WIDTH = tower_width
    Dims.SEGMENT_LENGTH = length / segments
    Dims.SEGMENTS = segments
    Dims.IS_CONNECTED = False

    create_segments()
    create_beams()


def create_connected(tower_nodes, tower_beams, tower_height, tower_width):
    """
    Creates jib attached to the tower as to have the entire crane in one block

    Args:
    :param tower_nodes: nodes which make up the tower (float[])
    :param tower_beams: beams which make up the tower (int[])
    :param length: length of the jib to be created  (float)
    :param segments: number of segments the jib should consist of (int)
    """
    global nodes
    nodes = tower_nodes
    global beams
    beams = tower_beams
    
    Dims.START_HEIGHT = tower_height #(np.asarray(nodes).astype(float))[len(nodes) - 1, 2]
    Dims.TOWER_WIDTH = tower_width
    Dims.IS_CONNECTED = True
    Dims.INIT_BAR = max(np.asarray(beams).astype(int).max() - 1, 0) # wrapped in max just in case

    create_segments()
    create_beams()


def create_segments():
    """Create the segments of a jib"""
    for i in range(Dims.SEGMENTS + 1):
        if not (i == 0 and Dims.IS_CONNECTED):  # skips the first run-through if nodes already exist
            nodes.append([Dims.TOWER_WIDTH + Dims.SEGMENT_LENGTH * i, 0, Dims.START_HEIGHT])
            nodes.append([Dims.TOWER_WIDTH + Dims.SEGMENT_LENGTH * i, Dims.TOWER_WIDTH, Dims.START_HEIGHT])
        if i < Dims.SEGMENTS:
            nodes.append([Dims.TOWER_WIDTH + Dims.SEGMENT_LENGTH * i + Dims.SEGMENT_LENGTH / 2, Dims.SEGMENT_LENGTH / 2,
                          Dims.START_HEIGHT + Dims.HEIGHT])


def create_beams():
    """Create all beams of a single segment"""
    for i in range(Dims.SEGMENTS):
        val_to_add = 3 * i + Dims.INIT_BAR
        create_horizontal_beams(i, val_to_add)
        create_diagonal_beams(val_to_add)


def create_horizontal_beams(i, val_to_add):
    """Create the horizontal beams of a segment"""
    if i == 0 and not Dims.IS_CONNECTED:
        append_beam(0 + val_to_add, 1 + val_to_add)  # first horizontal (0-1)
    if i < Dims.SEGMENTS - 1:
        append_beam(2 + val_to_add, 5 + val_to_add)  # top connection
    append_beam(1 + val_to_add, 4 + val_to_add)
    append_beam(4 + val_to_add, 3 + val_to_add)
    append_beam(3 + val_to_add, 0 + val_to_add)
    append_beam(1 + val_to_add, 3 + val_to_add)  # diagonal beam


def create_diagonal_beams(val_to_add):
    """Create the diagonal beams of a segment, here the diagonal beams are the side of a pyramid"""
    append_beam(0 + val_to_add, 2 + val_to_add)
    append_beam(1 + val_to_add, 2 + val_to_add)
    append_beam(4 + val_to_add, 2 + val_to_add)
    append_beam(3 + val_to_add, 2 + val_to_add)


def get_nodes():
    """Return the nodes of jib as numpy array of type float64"""
    return np.array(nodes).astype(float)


def get_nodes_raw():
    """Returns the nodes of the tower in original format"""
    return nodes


def get_beams():
    """Return the beams of jib as numpy array"""
    return np.array(beams)


def get_beams_raw():
    """Returns the beams of the tower in original format"""
    return beams


def append_beam(start_node, end_node):
    """
    Creates a beam between 2 given points and adds the length to a running total
    
    Args:
    :param start_node: start node of the beam
    :param end_node: end node of the beam
    """
    beams.append([start_node, end_node])
    start_float = np.array(nodes[start_node]).astype(float)
    end_float = np.array(nodes[end_node]).astype(float)
    Dims.TOTAL_LENGTH += np.linalg.norm(end_float - start_float)


def get_length():
    """Returns total length of all beams"""
    return Dims.TOTAL_LENGTH


def get_dims():
    length = 0
    while length < 500 or length > 10000: # 5000-10000
        length = float(input('Enter the length of the jib in mm: '))
    seg_length = 0
    segs = 0
    while seg_length < 500 or seg_length > 2000: # 500-2000
        segs = int(input('Enter the how many segments you would like: '))
        seg_length = length / segs
    height = 0
    while height < 500 or height > 2000:
        height = float(input('Enter the height of the truss in mm: '))
    if np.sqrt(height ** 2 + (1/2 * np.sqrt(seg_length ** 2 + Dims.TOWER_WIDTH ** 2)) ** 2) > 2000:
        print(f'Warning! The diagonal elements will have a length of {np.sqrt(height ** 2 + (1/2 * np.sqrt(seg_length ** 2 + Dims.TOWER_WIDTH ** 2)) ** 2):.3f}mm which is greater than the 2000mm allowed!')
        print('Please adjust the measurements and reenter them.')
        get_dims()
    
    Dims.SEGMENTS = segs
    Dims.SEGMENT_LENGTH = seg_length
    Dims.HEIGHT = height

File: src/truss/jib.cpp
#include "jib.h"
#include <iostream>
#include <vector>
#include <cmath>


class Comps {
public:
    std::vector<std::vector<double>> nodes;
    std::vector<std::vector<int>> beams;
};
Comps comps;


double Jib::getJibLength() {
    return length;
}

double Jib::getJibHeight() {
    return height;
}

double Jib::getJibSegments() {
    return numberOfSegments;
}

double Jib::getJibTotalBeamLength() {
    return totalLength;
}

double Jib::getJibLongestBeam() {
    return longestBeam;
}

std::vector<std::vector<double>> Jib::getJibNodes() {
    return comps.nodes;
}

std::vector<std::vector<int>> Jib::getJibBeams() {
    return comps.beams;
}

int Jib::getEndBase() {
    return comps.nodes.size();
}

void Jib::setJibLength(double length) {
    this->length = length;
}

void Jib::setJibHeight(double height) {
    this->height = height;
}

void Jib::setJibSegments(double numberOfSegments) {
    this->numberOfSegments = numberOfSegments;
}

void Jib::setDimensions(double length, double height, int numSegs, int supStyle,
                   bool dropdown, bool bend) {
    // Reset arrays
    comps.nodes.clear();
    comps.beams.clear();
    // Reset calculated dimensions
    totalLength = 0;
    longestBeam = 0;
    // Set remaining parameters
    this->height = height;
    this->length = length;
    numberOfSegments = numSegs;
    switch (supStyle) {
        case 1:
            this->supStyle = JibStyle::TRUSS;
            break;
        case 2:
            this->supStyle = JibStyle::SET_BACK_TRUSS;
            break;
        default:
            this->supStyle = JibStyle::NONE;
            std::cout << "No support style chosen";
    }
    this->dropdown = dropdown;
    this->bend = bend;
}


void Jib::createJib(std::vector<std::vector<double>> nodes, std::vector<std::vector<int>> beams,
                   double towerHeight, double towerWidth) {
    comps.nodes = nodes;
    comps.beams = beams;
    startHeight = towerHeight;
    this->towerWidth = towerWidth;

    createSegments();
    createBeams();
}

void Jib::createSegments() {
    for (int i = 0; i < numberOfSegments; i++) {
        double jibLength_m = numberOfSegments * length / 1000;
        double lenIn_m = towerWidth / 1000 + length / 1000 * i;
        double bendGrad = 0.01 * pow(30, 1/jibLength_m * lenIn_m) * 1000;
        double botHeight = bend ? (startHeight + bendGrad) : startHeight;
        // skips the first run-through if nodes already exist
        if (i != 0) {
            comps.nodes.push_back({towerWidth + length * i, 0, botHeight});
            comps.nodes.push_back({towerWidth + length * i, towerWidth, botHeight});
        }
        double topHeight = 0;
        if (dropdown) {
            if (i <= 3 * numberOfSegments / 7) {
                topHeight = height;
            } else if (i >= 5 * numberOfSegments / 7) {
                topHeight = 0.76 * height;
            } else {
                double count = i - 2 * numberOfSegments / 7;
                topHeight = height - count * (0.24 * 7 * height / (2 * jibLength_m));
                topHeight = std::max(topHeight, 0.76 * height);
            }
        } else {
            topHeight = height;
        }
        // adds top nodes
        if (i < numberOfSegments) { //supType 1 -> truss
            comps.nodes.push_back({supStyle == JibStyle::TRUSS ? towerWidth + length * i + length / 2 : towerWidth * 1.15 + length * i,
                                   length / 2, botHeight + topHeight});
        }
    }
}

void Jib::createBeams() {
    for (int i = 0; i < numberOfSegments; i++) {
        double valToAdd = 3 * i + initBeam;
        createHorizontalBeams(i, valToAdd);
        createDiagonalBeams(valToAdd);
    }
}

void Jib::createHorizontalBeams(int seg, double valToAdd) {
    // if (i == 0 and not Dims.IS_CONNECTED) {
    //     appendBeam(0 + valToAdd, 1 + valToAdd);  // first horizontal (0-1)
    // }
    if (seg < numberOfSegments - 1) {
        appendBeam(2 + valToAdd, 5 + valToAdd);  // top connection
    }
    // if (i == Dims.SEGMENTS - 1 and Dims.SUPPORT_TYPE != Style.TRUSS) {
    //     appendBeam(2 + valToAdd, 5 + valToAdd);
    // }
    appendBeam(1 + valToAdd, 4 + valToAdd);
    appendBeam(4 + valToAdd, 3 + valToAdd);
    appendBeam(3 + valToAdd, 0 + valToAdd);
    appendBeam(1 + valToAdd, 3 + valToAdd);     // bottom diagonal beam
}

void Jib::createDiagonalBeams(double valToAdd) {
    appendBeam(0 + valToAdd, 2 + valToAdd);
    appendBeam(1 + valToAdd, 2 + valToAdd);
    appendBeam(4 + valToAdd, 2 + valToAdd);
    appendBeam(3 + valToAdd, 2 + valToAdd);
}

void Jib::appendBeam(int startNode, int endNode) {
    // // Create a beam between the two given nodes
    comps.beams.push_back({startNode, endNode});

    // Calculate the length of the beam
    auto startVector = comps.nodes[startNode];
    auto endVector = comps.nodes[endNode];
    std::vector<double> lenVector = {endVector[0] - startVector[0], endVector[1] - startVector[1], endVector[2] - startVector[2]};
    double length = sqrt(pow(lenVector[0], 2) + pow(lenVector[1], 2) + pow(lenVector[2], 2));

    // Update the longest beam and total length
    longestBeam = std::max(length, longestBeam);
    totalLength += length;
}